Pin state transition service

supasate opened the pull request:
This PR is for issue #140 and depends on activity log service #142. 

The following is the checklist for this PR. Feel free to comment each commit while implementing this PR. 

Write log
- [x] Prepare log info in before hook and pass it to after hook
- [x] Add test: prepare-activity-log hook
- [x] Call activity log service in after hook
- [x] Add test: log-activity hook

Transition to `rejected` state
- [x] Handle transition 
  `POST /pins/:pinId/state_transition -H "Content-type: application/json" -H "Authorization: Bearer {token}" -d '{"state":"rejected"}'`
- [x] Add test: check pin state is changed as expected
- [x] Add test: allow transition from `unverified` to `rejected` 
- [x] Add test: allow transition from `verified` to `rejected` 
- [x] Add test: prevent transition from other states to `rejected`
- [x] Add test: allow only `super_admin` and `organization_admin` to do this operation 


Transition to `verified` state
- [x] Handle transition
  `POST /pins/:pinId/state_transition -H "Content-type: application/json" -H "Authorization: Bearer {token}" -d '{"state": "verified"}'`
- [x] Add test: check pin state is changed as expected
- [x] Add test: allow transition from `unverified` to `verified` 
- [x] Add test: allow transition from `assigned` to `verified`
- [x] Add test: prevent transition from other states to `verified`
- [x] Add test: allow only `super_admin`, `organization_admin`, and `department_head` to do this operation 

Transition to `unverified` state
- [x] Handle transition
  `POST /pins/:pinId/state_transition -H "Content-type: application/json" -H "Authorization: Bearer {token}" -d '{"state": "unverified"}'`
- [x] Add test: check pin state is changed as expected
- [x] Add test: allow transition from `verified` to `unverified` 
- [x] Add test: prevent transition from other states to `unverified`
- [x] Add test: allow only `super_admin` and `organization_admin` to do this operation 

Transition to `assigned` state
- [x] Handle transition
  `POST /pins/:pinId/state_transition -H "Content-type: application/json" -H "Authorization: Bearer {token}" -d '{"state": "assigned", "assigned_department":"{departmentId}"}'`
- [x] Add test: check pin state is changed as expected
- [x] Add test: allow transition from `unverified` to `assigned` 
- [x] Add test: allow transition from `verified` to `assigned`
- [x] Add test: prevent transition from other states to `assigned`
- [x] Add test: allow only `super_admin` and `organization_admin` to do this operation 


Transition to `processing` state
- [x] Handle transition 
  `POST /pins/:pinId/state_transition -H "Content-type: application/json" -H "Authorization: Bearer {token}" -d '{"state": "processing", "processed_by":"{userId}"}'`
- [x] Add test: check pin state is changed as expected
- [x] Add test: allow transition from `assigned` to `processing`
- [x] Add test: allow transition from `resolved` to `processing`
- [x] Add test: prevent transition from other states to `processing`
- [x] Add test: allow only `super_admin`, `organization_admin`, and `department_head` to do this operation 

Transition to `resolved` state
- [x] Handle transition
  `POST /pins/:pinId/state_transition -H "Content-type: application/json" -H "Authorization: Bearer {token}" -d '{"state": "resolved"}'`
- [x] Add test: check pin state is changed as expected
- [x] Add test: allow transition from `processing` to `resolved` 
- [x] Add test: prevent transition from other states to `resolved`
- [x] Add test: allow only `department_head` to do this operation

supasate committed:
Change User model to contain an array of organization-department pairs

supasate committed:
Change Pin model to contain archived flag and assignment

supasate committed:
Create available pin states

supasate committed:
Create before and after hook for state-transition logging

supasate committed:
Create state transition service - still handle only reject action

supasate committed:
Separate prepare-activity-log hook and log-activity hook into new files

supasate committed:
Add test for log-activity hook

supasate committed:
[Fix] convert ObjectID to String before comparing

supasate committed:
Add test for prepare-activity-log hook

supasate committed:
Prepare fixtures for users in each role

supasate committed:
Check valid state transition by previous state, next state, and role

supasate committed:
Add processed_by field for processing state

supasate committed:
Fix array.append bug and add safeguard to the hook

supasate committed:
Update pin properties for each state transition

supasate committed:
Add test 'unverified' state transition

supasate committed:
Add test: 'verified' state transition

supasate committed:
Add test: 'assigned' state transition

supasate committed:
Add test: 'processing' state transition

supasate committed:
Add test: 'resolved' state transition

supasate commented:
@parnurzeal All done for state transition actions.

supasate committed:
Rename path or activity-log service into plural form

parnurzeal commented on test/fixtures/organizations.js:
```diff
@@ -2,7 +2,8 @@ module.exports = [
     _id: '57933111556362511181aaa1',
     name: 'YouPin',
-    users: ['579334c74443625d6281b6ff', '579334c75563625d6281b6f1'], // superAdminUser and adminUser
+    // superAdminUser, organizationAdminUser, departmentHeadUser
```
I think let's change these numbers to constants for testing in later PR. Too hard to figure it out what they are in other places.

supasate replied:
Totally agree!

parnurzeal commented on src/services/pin-state-transition/hooks/prepare-activity-log.js:
```diff
@@ -0,0 +1,140 @@
+  }
+
+  // hook.data.state should be provided via body data
+  if (!hook.data.state) {
```
Can `hook.data` be undefined or null?

supasate replied:
For `create` method, `hook.data` is always an object. (https://github.com/feathersjs/feathers-commons/blob/8e6b85681a2f1c6f8193c781cddb8086c21fba6d/src/hooks.js#L27-L33)

parnurzeal commented on test/fixtures/organization_admin_user.js:
```diff
@@ -8,5 +8,11 @@ module.exports = {
+  organization_and_role_pairs: [
+    { organization: '579334c74443625d6281b6dd', role: ORGANIZATION_ADMIN },
+  ],
+  organization_and_department_pairs: [
```
Sounds like this guy is an admin for all departments. How do we manage this kind of usecase? Do we need to add all departments in this organization the list?

supasate replied:
I don't think so. This guy can be only belong to his department but has been assigned the `organization_admin` role to take care of the admin task.

parnurzeal replied:
then... shouldn't it be deparment_admin?

supasate replied:
Oh, got it. That org_admin guy must be in `57933111556362511181ccc1` not `57933111556362511181bbb1`.

parnurzeal replied:
No, I mean if he is the organization admin, shouldn't he be able to do whatever to any department in his organization?

supasate replied:
Depends on organization policy? IMO, for the faculty of engineering, the responsibility of the Saraban team (organization admin) is screening and assigning tasks to other departments.

parnurzeal replied:
ok LGTM. Let's squash it and merge.

parnurzeal commented on src/services/pin-state-transition/hooks/prepare-activity-log.js:
```diff
@@ -0,0 +1,140 @@
+  safetyCheck(hook);
+
+  const pinId = hook.params.pinId;
+  const user = hook.params.user.name;
```
s/user/nameOfUser/

supasate replied:
Okay

parnurzeal commented on src/services/pin-state-transition/hooks/prepare-activity-log.js:
```diff
@@ -0,0 +1,140 @@
+        if (previousState === states.UNVERIFIED) {
+          action = actions.VERIFY;
+          description = `${user} verified pin ${pin._id}`;
+        } else if (previousState === states.ASSIGNED) {
```
I don't understand the logic here. Why does (prev==ASSIGNED && next==VERIFIED) cause the DENY action?

supasate replied:
A pin was assigned to a department but the department denied that pin. So, the pin will go back to `verified` but not `assigned` state (wait for re-assign to new department).

parnurzeal replied:
seems not too intuitive... but okay.
Can you comment more about this?

supasate replied:
Done

parnurzeal commented on src/services/pin-state-transition/hooks/prepare-activity-log.js:
```diff
@@ -0,0 +1,140 @@
+        changedFields.push('assigned_department');
+        previousValues.push(pin.assigned_department);
+        updatedValues.push(hook.data.assigned_department);
+        action = actions.ASSIGN;
```
put action on top like above actions you did.

supasate replied:
Okay

parnurzeal commented on src/services/pin-state-transition/hooks/prepare-activity-log.js:
```diff
@@ -0,0 +1,140 @@
+        changedFields.push('processed_by');
+        previousValues.push(pin.processed_by);
+        updatedValues.push(hook.data.processed_by);
+        action = actions.PROCESS;
```
put action on top like above actions you did.

supasate replied:
Okay

parnurzeal commented on src/services/pin-state-transition/hooks/prepare-activity-log.js:
```diff
@@ -0,0 +1,140 @@
+      case states.PROCESSING:
+        changedFields.push('processed_by');
+        previousValues.push(pin.processed_by);
+        updatedValues.push(hook.data.processed_by);
```
Should safetyCheck also check this is not null in `hook.data` for this action?

supasate replied:
Okay

parnurzeal commented on src/services/pin-state-transition/hooks/prepare-activity-log.js:
```diff
@@ -0,0 +1,140 @@
+          // If a department denies assignment, remove assigned_department value
+          action = actions.DENY;
+          changedFields.push('assigned_department');
+          previousValues.push(pin.assigned_department);
```
Should safetyCheck also check this is not null in `hook.data` for this action?

supasate replied:
Can't.

`pin.assiged_department` is a property of a pin in the database not a property of the hook object. So, we just push the value that we get from the database. It can be `null` or `undefined` but we don't care. We just keep record of its previous value.

parnurzeal commented on src/services/pin-state-transition/hooks/prepare-activity-log.js:
```diff
@@ -0,0 +1,140 @@
+      organization: pin.organization,
+      department: departments,
+      actionType: actions.types.STATE_TRANSITION,
+      action,
```
Wow, I just knew that we can just add variable like this w/o using colon. This only works if the variable name is the same with the field name?

supasate replied:
Yes. It's ES6 shorthand syntax (and eslint force us to do so).

parnurzeal commented on src/services/pin-state-transition/index.js:
```diff
@@ -0,0 +1,136 @@
+
+  static isValidStateTransition(prevState, nextState, role) {
+    // Map previous state and role to a list of possible next states
+    const possibleNextStates = {
```
I feel it would be more intuitive to use Action instead of States.
Like `Person1 can do Action1` instead of `Person1 do sth. then you get State1`.

But I can see it is more complicated to use Action looking from the rest of the code structured for nextState. This is good enough.

supasate replied:
I see. I tried to map with the RESTful API that specifies **what** the shape of resource we want it to be, not **how**. I still feel uncomfortable with it either or I don't fully understand it yet. Let it be for now.

parnurzeal commented on src/services/pin-state-transition/index.js:
```diff
@@ -0,0 +1,136 @@
+
+    // Need additional property for ASSIGNED and PROCESSING states
+    if (nextState === ASSIGNED) {
+      if (!data.assigned_department) {
```
I see now. If assigned_department is null, it seems prepare-log-activity push null before getting to this state. But then if it is null, it will stop here and the logInfo will not get to use and log. Is it correct?

supasate replied:
Correct. 

However, I will add safety check to both files. So, it can work separately from each other in the future.

supasate committed:
1. Add more safety checks
2. Rename a variable to be more understandable
3. Fix department of organization admin in fixture
4. Add more comment about unintuitive state transition
